src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -67,69 +67,128 @@ public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentCont
 
             FeatureVariant variant = null;
 
-            double cumulativePercentage = 0;
+            var lookup = new Dictionary<FeatureVariant, TargetingFilterSettings>();
+
+            //
+            // Check users
+            foreach (FeatureVariant v in featureDefinition.Variants)
+            {
+                TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+
+                //
+                // Put in lookup table to avoid repeatedly creating targeting settings
+                lookup[v] = targetingSettings;
+
+                if (targetingSettings == null &&
+                    v.Default)
+                {
+                    //
+                    // Valid to omit audience for default variant
+                    continue;
+                }
+
+                if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                {
+                    throw new ArgumentException(reason, paramName);
+                }
+
+                //
+                // Check if the user is being targeted directly
+                if (targetingSettings.Audience.Users != null &&
+                    TargetingEvaluator.IsTargeted(
+                        targetingContext,
+                        targetingSettings.Audience.Users,
+                        _options.IgnoreCase))
+                {
+                    return new ValueTask<FeatureVariant>(v);
+                }
+            }
 
             var cumulativeGroups = new Dictionary<string, double>(
                 _options.IgnoreCase ? StringComparer.OrdinalIgnoreCase :
                                       StringComparer.Ordinal);
 
+            //
+            // Check Groups
             foreach (FeatureVariant v in featureDefinition.Variants)
             {
-                TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+                TargetingFilterSettings targetingSettings = lookup[v];
 
-                if (targetingSettings == null)
+                if (targetingSettings == null ||
+                    targetingSettings.Audience.Groups == null)
                 {
-                    if (v.Default)
-                    {
-                        //
-                        // Valid to omit audience for default variant
-                        continue;
-                    }
+                    continue;
                 }
 
-                if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                AccumulateGroups(targetingSettings.Audience, cumulativeGroups);
```

Comment:
>targetingSettings.Audience [](http://example.com/codeflow?start=33&length=26)

Should we pass `targetingSettings.Audience.Groups` now that this method specifically accumulates groups?

Reply:
I think that's a great idea.